### PR TITLE
feat: phoenixd wallet integration

### PR DESCRIPTION
**Pull request by bitkarrot**
see https://github.com/lnbits/lnbits/issues/2350 for reference

### CODE REVIEW

**Comment by prusnak** [comment]
Don't touch the description in lnbits/static/i18n - we are going to simplify it in https://github.com/lnbits/lnbits/pull/2356 - it is pain in the ass to keep it up to date.

**Comment by bitkarrot** [comment]
@prusnak should i roll the changes for i18n back on this PR?

> Don't touch the description in lnbits/static/i18n - we are going to simplify it in #2356 - it is pain in the ass to keep it up to date.

**Comment by bitkarrot** [comment]
Status update: Only items left to do on this PR: 

- get the websocket working for graphql (not sure how much of a requirements this might be) 
- unit test. 

everything else send/receive is working

**Comment by bitkarrot** [comment]
- websocket now working
- TODO: unit test. 
- what gives with this receive payment and need refresh to see update?

**Comment by prusnak** [comment]
> @prusnak should i roll the changes for i18n back on this PR?

PR has been merged

Just rebase on top of the current dev.

**Comment by bitkarrot** [comment]
@prusnak i rebased, selected all upstream "--theirs" for i18n, but on git it still thinks i have conflicts on the i18n, not sure why. 

This PR now marked ready for review

**Comment by bitkarrot** [comment]
Allow edits has been checked now, is ready for review

**Comment by bitkarrot** [comment]
Thank you @dni for helping me fix the messy rebase!

**Comment by prusnak** [comment]
Rebased again to fix conflicts

**Review by prusnak — APPROVED** [review]
utACK

File: lnbits/wallets/phoenixd.py [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+            return PaymentResponse(False, None, None, None, error_message)
+
+        data = r.json()
+        logger.info(f"pay_invoice data: {data}")
+
+        checking_id = data["paymentHash"]
+        fee_msat = -int(data["routingFeeSat"])
+        preimage = data["paymentPreimage"]
+
+        return PaymentResponse(True, checking_id, fee_msat, preimage, None)
```
Comment (callebtc) [inline lnbits/wallets/phoenixd.py:129]: This is very dangerous. `True` should never be returned directly but only through a dictionary that explicitly checks for a valid response from the backend (see other wallet files).

File: lnbits/wallets/phoenixd.py [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+        if r.is_error:
+            return PaymentPendingStatus()
+        data = r.json()
+        # logger.info(f'get_invoice_status data: {data}')
+
+        fee_msat = data["fees"]
+        preimage = data["preimage"]
+        is_paid = data["isPaid"]
+
+        return PaymentStatus(paid=is_paid, fee_msat=fee_msat, preimage=preimage)
```
Comment (callebtc) [inline lnbits/wallets/phoenixd.py:142]: This is a good example of how `paid` should be set. Less room for error.

File: lnbits/wallets/phoenixd.py [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+                "cannot initialize PhoenixdWallet: missing phoenixd_api_endpoint"
+            )
+        if not settings.phoenixd_api_password:
+            raise ValueError(
+                "cannot initialize PhoenixdWallet: missing phoenixd_api_password"
+            )
+
+        self.endpoint = self.normalize_endpoint(settings.phoenixd_api_endpoint)
+
+        self.ws_url = f"ws://{urllib.parse.urlsplit(self.endpoint).netloc}/websocket"
```
Comment (motorina0) [inline lnbits/wallets/phoenixd.py:39]: Why is the protocol `ws://` and not `wss://`?

Comment (bitkarrot) [reply lnbits/wallets/phoenixd.py:39]: its like the cliche endpoint, phoenixd is also just running http. not https

File: lnbits/wallets/phoenixd.py (hunk 2 above)
Comment (bitkarrot) [reply lnbits/wallets/phoenixd.py:142]: ok

File: lnbits/wallets/phoenixd.py (hunk 1 above)
Comment (bitkarrot) [reply lnbits/wallets/phoenixd.py:129]: This `True` is returned directly in https://github.com/lnbits/lnbits/blob/4f118c5f98247dce8509089a8c3660099df2bff3/lnbits/wallets/lntips.py#L122 and https://github.com/lnbits/lnbits/blob/4f118c5f98247dce8509089a8c3660099df2bff3/lnbits/wallets/opennode.py#L115 and https://github.com/lnbits/lnbits/blob/4f118c5f98247dce8509089a8c3660099df2bff3/lnbits/wallets/lnpay.py#L127 and https://github.com/lnbits/lnbits/blob/4f118c5f98247dce8509089a8c3660099df2bff3/lnbits/wallets/zbd.py#L120 and possibly a few other backend wallet implementations.  🧐

Please specify which wallet backend should be the correct model, there's at least 4 which are currently setup

@prusnak @motorina0

**Comment by arbadacarbaYK** [comment]
Pls also update the guide with the credentials on merge

File: lnbits/wallets/phoenixd.py (hunk 1 above)
Comment (bitkarrot) [reply lnbits/wallets/phoenixd.py:132]: @motorina0 @dni 

bump. what should i do?

**Comment by bitkarrot** [comment]
> Pls also update the guide with the credentials on merge

do you mean this? https://github.com/lnbits/lnbits/pull/2362/files#diff-382e155c2929164da086221c0aad99285747326757ed3a47854faef4d211b3c3

**Comment by arbadacarbaYK** [comment]
> > Pls also update the guide with the credentials on merge
> 
> do you mean this? https://github.com/lnbits/lnbits/pull/2362/files#diff-382e155c2929164da086221c0aad99285747326757ed3a47854faef4d211b3c3

Yes, thanks 🚀🚀

**Comment by motorina0** [comment]
Added unit tests. Updated code to pass tests.
Observations:
 - use `headers` field for [client initialisation](https://github.com/lnbits/lnbits/pull/2362/commits/b42f9827aa7f9c1b287c18407072b2e5836db438#diff-dd5221a5b7e9defe1cbdc1f3e3e0d709cdea670356f838088ee785950af45bedR49). @bitkarrot let me know if there is a reason not to use directly the `headers` field
 - surround methods with `try-catch`
 - fixed `PaymentStatus` not to incorrectly signal failure

Todo:
 - `get_payment_status` should use the outgoing endpoint (see comment)

Note:
 - unit tests do not cover `paid_invoices_stream`

File: lnbits/wallets/phoenixd.py (hunk 1 above)
Comment (motorina0) [reply lnbits/wallets/phoenixd.py]: > This is very dangerous. `True` should never be returned directly but only through a dictionary that explicitly checks for a valid response from the backend (see other wallet files).

The [endpoint](https://phoenix.acinq.co/server/api#pay-invoice) does not return a `"success"` kind of flag (unfortunately):

**Response**
```json
{
    "recipientAmountSat": 1,
    "routingFeeSat": 4,
    "paymentId": "961d413c-91ca-4f3d-9d91-760a08405901",
    "paymentHash": "9586b2819f3c6d3c3e087984b97dcf4b74babfcaa5da4bc4207f47c6c1cc71e0",
    "paymentPreimage": "6a3e65d9a05113864c9c27e430cc99751359ccbfa8fd5869aea46405df657986"
}
```

The check has to rely on the mandatory fields being present (like `"paymentHash"`).

Given the API, there isn't anything else to be done.

**Review by motorina0 — APPROVED** [review]

**Comment by motorina0** [comment]
Regarding the `LSP fee` for incoming payments: 
  - the `lnbits user` generates the invoice
  - the payer pays the Lightning Network fee
  - extra: the `lnbits user` also pays a fee to Phonenix

In the below screenshot I have hardcoded a fee of `1500 msat`. The fee computation for incoming payments is correct.
**Incoming**: 110 (two payments, one of `100 sats` and one of `10 sats`)
**Fees**: `3 sats` (2 x `1500 msat`)
**Wallet Balance**: `107 sat` (`100 - 3`)

image

**Review by dni — APPROVED** [review]

**Comment by Funman2** [comment]
would be great to have a docker image!

**Comment by motorina0** [comment]
> would be great to have a docker image!

you can find the LNbits docker images here:
https://hub.docker.com/r/lnbits/lnbits/tags

**Pull request merged**